### PR TITLE
Add mosaic and horizontal

### PR DESCRIPTION
**Pull request by SungjunApp**
ticket: https://pixlee.atlassian.net/browse/IMP-1247

## Background 
- Stradivarius needs a mosaic and horizontal list
- This is to implement mosaic view and horizontal view that are the same features of this [Android PR](https://github.com/pixlee/android-sdk/pull/52) which is already released

## Implementation
- changed the minimum iOS version from 10 to 13: 
   - [checkout the ios distribution status](https://developer.apple.com/support/app-store/)
- Currently, PXLWidgetView in production supports 
  - Grid: displaying multiple items in a row
  - List: displaying one item in a row
- additionally, this pr adds
  - Mosaic: similar to our Mosaic on Widget
  - Horizontal: a horizontal version of Mosaic
- updated Fastlane libraries
### Mosaic
- `mosaicSpan` provides three span options, [three, four, five].
#### mosaicSpan: .three
https://user-images.githubusercontent.com/6112156/170452344-64bc2361-6469-4910-8cf6-6b38713fd690.mov

#### mosaicSpan: .four
https://user-images.githubusercontent.com/6112156/170450707-e932bbcd-247c-4278-b097-629bf317c8e3.mov

#### mosaicSpan: .five
https://user-images.githubusercontent.com/6112156/170450717-36d28c9c-a318-4f49-bd47-c7620345efe1.mov

### Horizontal
https://user-images.githubusercontent.com/6112156/170450730-057e0925-ec2d-4772-8987-3078de104479.mov

#### Mosaic code snippet
- `mosaicSpan`: provides three span options, [three, four, five].
- `cellPadding`: space between cells
```swift
extension WidgetViewController: PXLWidgetViewDelegate {
    func setWidgetSpec() -> WidgetSpec {
        WidgetSpec.horizontal(
            .init(
                mosaicSpan: .five,
                cellPadding: 1,
                loadMore: .init(cellHeight: 100.0,
                                cellPadding: 10.0,
                                text: "LoadMore",
                                textColor: UIColor.darkGray,
                                textFont: UIFont.systemFont(ofSize: UIFont.buttonFontSize),
                                loadingStyle: .gray)))
    }
}
```
```swift
class WidgetViewController: UIViewController {
    var widgetView = PXLWidgetView()
    override func viewDidLoad() {
        super.viewDidLoad()
        widgetView.delegate = self
        view.addSubview(widgetView)

        if let pixleeCredentials = try? PixleeCredentials.create() {
            let albumId = pixleeCredentials.albumId
            let album = PXLAlbum(identifier: albumId)
            album.filterOptions = PXLAlbumFilterOptions(hasPermission: true)
            album.sortOptions = PXLAlbumSortOptions(sortType: .approvedTime, ascending: false)
            album.perPage = 18
            widgetView.searchingAlbum = album
        }
    }
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        // make the view full screen
        widgetView.frame = CGRect(x: 0, y: 0, width: view.frame.size.width, height: view.frame.size.height)
    }
}
```
#### Horizontal code snippet
- `cellPadding`: space between cells
```swift
extension WidgetViewController: PXLWidgetViewDelegate {
    func setWidgetSpec() -> WidgetSpec {
        WidgetSpec.horizontal(
            .init(
                cellPadding: 1,
                loadMore: .init(cellHeight: 100.0,
                                cellPadding: 10.0,
                                text: "LoadMore",
                                textColor: UIColor.darkGray,
                                textFont: UIFont.systemFont(ofSize: UIFont.buttonFontSize),
                                loadingStyle: .gray)))
    }
}
```
```swift
class WidgetViewController: UIViewController {
    var widgetView = PXLWidgetView()
    override func viewDidLoad() {
        super.viewDidLoad()
        widgetView.delegate = self
        view.addSubview(widgetView)

        if let pixleeCredentials = try? PixleeCredentials.create() {
            let albumId = pixleeCredentials.albumId
            let album = PXLAlbum(identifier: albumId)
            album.filterOptions = PXLAlbumFilterOptions(hasPermission: true)
            album.sortOptions = PXLAlbumSortOptions(sortType: .approvedTime, ascending: false)
            album.perPage = 18
            widgetView.searchingAlbum = album
        }
    }
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        // set the height
        widgetView.frame = CGRect(x: 0, y: 0, width: view.frame.size.width, height: view.frame.size.height/5)
    }
}
```
#### Full Example
```swift
import Foundation
import PixleeSDK
import UIKit

class WidgetViewController: UIViewController {
    static func getInstance() -> WidgetViewController {
        let vc = WidgetViewController(nibName: "EmptyViewController", bundle: Bundle.main)
        return vc
    }

    var widgetView = PXLWidgetView()

    override func viewDidLoad() {
        super.viewDidLoad()
        widgetView.delegate = self
        view.addSubview(widgetView)

        if let pixleeCredentials = try? PixleeCredentials.create() {
            let albumId = pixleeCredentials.albumId
            let album = PXLAlbum(identifier: albumId)
            album.filterOptions = PXLAlbumFilterOptions(hasPermission: true)
            album.sortOptions = PXLAlbumSortOptions(sortType: .approvedTime, ascending: false)
            album.perPage = 18
            widgetView.searchingAlbum = album
        }
    }

    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        widgetView.frame = CGRect(x: 0, y: 0, width: view.frame.size.width, height: view.frame.size.height/5)
    }

    
}

// MARK: - Photo's click-event listeners
extension WidgetViewController: PXLPhotoViewDelegate {
    public func onPhotoButtonClicked(photo: PXLPhoto) {
        print("Action tapped \(photo.id)")
        openPhotoProduct(photo: photo)
    }

    public func onPhotoClicked(photo: PXLPhoto) {
        print("Photo Clicked \(photo.id)")
        openPhotoProduct(photo: photo)
    }

    func openPhotoProduct(photo: PXLPhoto) {
        present(PhotoProductListDemoViewController.getInstance(photo), animated: false, completion: nil)
    }
}

// MARK: Widget's UI settings and scroll events
extension WidgetViewController: PXLWidgetViewDelegate {
    func setWidgetSpec() -> WidgetSpec {
        // An example of Mosaic
//        WidgetSpec.mosaic(
//            .init(
//                mosaicSpan: .five,
//                cellPadding: 1,
//                loadMore: .init(cellHeight: 100.0,
//                                cellPadding: 10.0,
//                                text: "LoadMore",
//                                textColor: UIColor.darkGray,
//                                textFont: UIFont.systemFont(ofSize: UIFont.buttonFontSize),
//                                loadingStyle: .gray)))
        
        // An example of Mosaic
        WidgetSpec.horizontal(
            .init(
                cellPadding: 1,
                loadMore: .init(cellHeight: 100.0,
                                cellPadding: 10.0,
                                text: "LoadMore",
                                textColor: UIColor.darkGray,
                                textFont: UIFont.systemFont(ofSize: UIFont.buttonFontSize),
                                loadingStyle: .gray)))
    }

    func setWidgetType() -> String {
        "replace_this_with_yours"
    }

    func setupPhotoCell(cell: PXLGridViewCell, photo: PXLPhoto) {
        
        cell.setupCell(photo: photo, title: nil, subtitle: nil, buttonTitle: nil, configuration: PXLPhotoViewConfiguration(cropMode: .centerFill), delegate: self)
    }

    func scrollViewDidScroll(_ scrollView: UIScrollView) {
         print("scrollViewDidScroll \(scrollView)")
    }
}
```

### CODE REVIEW

**Comment by SungjunApp** [comment]
there's errors in building SDKs. I'll fix it tomorrow.

**Comment by SungjunApp** [comment]
Fixed all build errors. it's ready to be reviewed.

File: Gemfile.lock [hunk 1]
```diff
@@ -8,20 +8,20 @@ GEM
     artifactory (3.0.15)
     atomos (0.1.3)
     aws-eventstream (1.2.0)
-    aws-partitions (1.548.0)
-    aws-sdk-core (3.125.3)
+    aws-partitions (1.592.0)
```
Comment (jchen123) [inline Gemfile.lock:11]: wait I just noticed our ios SDK has a gemfile?  Why?

File: Sources/PixleeSDK/Widgets/Utilities/MosaicLayoutUtil.swift [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        originalSizeItem.contentInsets = NSDirectionalEdgeInsets(top: cellPadding, leading: cellPadding, bottom: cellPadding, trailing: cellPadding)
+        
+        return [
+            MosaicInfo(heightRatio: 1 , collectionLayoutGroup: NSCollectionLayoutGroup.horizontal(
+                layoutSize: NSCollectionLayoutSize(widthDimension: .fractionalHeight(1), heightDimension: .fractionalHeight(1)),
+                subitems: [originalSizeItem])),
+        ]
+    }
+    
+    func get3SpanList(cellPadding: CGFloat) -> [MosaicInfo] {
```
Comment (jchen123) [inline Sources/PixleeSDK/Widgets/Utilities/MosaicLayoutUtil.swift:115]: this function is pretty crazy. is this all copy pasted from a library or something or can it be explained?

File: Gemfile.lock (hunk 1 above)
Comment (SungjunApp) [reply Gemfile.lock:11]: Testing SDK is done on circleci with Fastlane. [Fastlane uses Gemfile](https://docs.fastlane.tools/getting-started/ios/setup/)

**Review by jchen123 — APPROVED** [review]

File: Sources/PixleeSDK/Widgets/Utilities/MosaicLayoutUtil.swift (hunk 2 above)
Comment (jchen123) [reply Sources/PixleeSDK/Widgets/Utilities/MosaicLayoutUtil.swift:115]: oh wow...

will this work for all phone sizes?

Comment (jharrilim) [reply Sources/PixleeSDK/Widgets/Utilities/MosaicLayoutUtil.swift:115]: I can't comment on how this should be done dynamically within the realms of iOS, but for static variants it seems good to me. Doesn't look like you can do much refactoring here tbh, all of the constructors just have a lot of parameters

Comment (SungjunApp) [reply Sources/PixleeSDK/Widgets/Utilities/MosaicLayoutUtil.swift:115]: @jchen123 Yes. This's how this works. The developer can decide how many spans among 3, 4, and 5 their app uses when using `PXLWidgetViewDelegate` via `WidgetSpec.mosaic()`. the rectangles stretch or shrink depending on the phone size. 

@jharrilim There're some possible patterns I haven't implemented. I have a theory of implementing this dynamically but am not sure how quickly I can implement this without causing bugs. So, I just took these static variants. Definitely, I'll soon take the dynamic approach to accommodate more patterns.

**Pull request merged**